utils/s2n_fork_detection.c

```diff
@@ -0,0 +1,351 @@
+
+static inline int initialise_inherit_zero(void *addr, long page_size)
+{
+#if defined(USE_MINHERIT) && defined(MAP_INHERIT_ZERO)
+    POSIX_ENSURE(minherit(addr, pagesize, MAP_INHERIT_ZERO) == 0, S2N_ERR_FORK_DETECTION_INIT);
+#endif
+
+    return S2N_SUCCESS;
+}
+
+static void pthread_atfork_on_fork(void)
+{
+  /* This zeroises the first byte of the memory page pointed to by
+   * *zero_on_fork_addr. This is the same byte used as fork event detection
+   * sentinel in s2n_get_fork_generation_number(). The same memory page, and in
+   * turn, the byte, is also the memory zeroised by the MADV_WIPEONFORK fork
+   * detection mechanism.
+   *
+   * Aquire locks to be on the safe side. We want to avoid the checks in
+   * s2n_get_fork_generation_number() getting executed before setting the sentinel
+   * flag. The write lock prevents any other thread from owning any other type
+   * of lock.
+   *
+   * pthread_atfork_on_fork() cannot return errors. Hence, there is no way to
+   * gracefully recover if [un]locking fails.
+   */
+  if (pthread_rwlock_wrlock(&fgn_state.fork_detection_rw_lock) != 0) {
+    abort();
+  }
+  *fgn_state.zero_on_fork_addr = 0;
+  if (pthread_rwlock_unlock(&fgn_state.fork_detection_rw_lock) != 0) {
+    abort();
+  }
+}
+
+static int inititalise_pthread_atfork(void)
+{
+  /* Register the fork handler pthread_atfork_on_fork that is excuted in the
+   * child process after a fork.
+   */
+  POSIX_ENSURE(pthread_atfork(NULL, NULL, pthread_atfork_on_fork) == 0, S2N_ERR_FORK_DETECTION_INIT);
+
+  return S2N_SUCCESS;
+}
+
+static void initialise_fork_detection_methods(void)
+{
+    void *addr = MAP_FAILED;
+    long page_size = 0;
+
+    /* Only used to disable fork detection mechanisms during testing. */
+    if (ignore_wipeonfork_or_inherit_zero_method_for_testing == S2N_FORK_DETECT_IGNORE &&
+        ignore_pthread_atfork_method_for_testing == S2N_FORK_DETECT_IGNORE) {
+
+        ignore_fork_detection_for_testing = S2N_FORK_DETECT_IGNORE;
+        goto end;
+    }
+
+    page_size = sysconf(_SC_PAGESIZE);
+    if (page_size <= 0) {
+        goto end;
+    }
+
+    addr = mmap(NULL, (size_t) page_size, PROT_READ | PROT_WRITE,
+                MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (addr == MAP_FAILED) {
+        goto end;
+    }
+
+    /* Some system don't define MADV_WIPEONFORK in sys/mman.h but the kernel
+     * still supports the mechanism (AL2 being a prime example). Likely because
+     * glibc on the system is old. We might be able to include kernel header
+     * files directly, that define MADV_WIPEONFORK, conditioning on specific
+     * OS's. But it is a mess. A more reliable method is to probe the system, at
+     * run-time, whether madvise supports the MADV_WIPEONFORK advice. However,
+     * the method to probe for this feature is equivalent to actually attempting
+     * to initialise the MADV_WIPEONFORK fork detection. Compare
+     * FOR_TESTING_probe_madv_wipeonfork_support() (used for testing) with
```

Comment:
This sentence does not make sense since function names has changed. Will fix in next push.